sonar-plugin-api/src/main/java/org/sonar/api/utils/ZipUtils.java

```diff
@@ -59,38 +62,88 @@ public static File unzip(File zip, File toDir) throws IOException {
     return unzip(zip, toDir, ze -> true);
   }
 
+  /**
+   * Unzip a file input stream to a directory.
+   *
+   * @param zip the zip file input stream
+   * @param toDir  the target directory. It is created if needed.
+   * @return the parameter {@code toDir}
+   * @since 6.2
+   */
   public static File unzip(InputStream zip, File toDir) throws IOException {
-    return unzip(zip, toDir, ze -> true);
+    return unzip(zip, toDir, null, ze -> true);
+  }
+
+  /**
+   * Unzip a file input stream to a directory.
+   *
+   * @param zip the zip file input stream
+   * @param toDir  the target directory. It is created if needed.
+   * @param unzipSizeThreshold  The parameter to prevent unzip size to exceed threshold(in Bytes)
+   * @return the parameter {@code toDir}
+   * @since 6.2
+   */
+  public static File unzip(InputStream zip, File toDir, long unzipSizeThreshold) throws IOException {
+    return unzip(zip, toDir, unzipSizeThreshold, ze -> true);
   }
 
   /**
    * Unzip a file to a directory.
    *
    * @param stream the zip input file
    * @param toDir  the target directory. It is created if needed.
+   * @param unzipSizeThreshold optional parameter to prevent unzip size to exceed threshold(in Bytes)
    * @param filter filter zip entries so that only a subset of directories/files can be
    *               extracted to target directory.
    * @return the parameter {@code toDir}
    * @since 6.2
    */
-  public static File unzip(InputStream stream, File toDir, Predicate<ZipEntry> filter) throws IOException {
+  public static File unzip(InputStream stream, File toDir, @Nullable Long unzipSizeThreshold, Predicate<ZipEntry> filter) throws IOException {
     if (!toDir.exists()) {
       FileUtils.forceMkdir(toDir);
     }
 
+    long totalSizeArchive = 0;
+
     Path targetDirNormalizedPath = toDir.toPath().normalize();
     try (ZipInputStream zipStream = new ZipInputStream(stream)) {
       ZipEntry entry;
       while ((entry = zipStream.getNextEntry()) != null) {
         if (filter.test(entry)) {
-          unzipEntry(entry, zipStream, targetDirNormalizedPath);
+          File target = getTargetFile(entry, targetDirNormalizedPath);
+          if (target.isDirectory()) {
+            continue;
+          }
+          if (unzipSizeThreshold != null) {
+            totalSizeArchive = extractZipEntry(zipStream, target, totalSizeArchive, unzipSizeThreshold);
+            if (totalSizeArchive > unzipSizeThreshold) {
```

Comment:
I think this can be removed if we move raising the exception to `extractZipEntry`

Reply:
good catch.